### PR TITLE
Add transforms document

### PR DESCRIPTION
**Pull request by aditidugar-zz**
Add documentation about supported data transforms in ML.NET.

@GalOshri please review categorization and help edit definitions.

[Internal Review URL](https://review.docs.microsoft.com/en-us/dotnet/machine-learning/resources/transforms?branch=pr-en-us-5752)

### CODE REVIEW

File: docs/machine-learning/resources/transforms.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+> [!NOTE]
+> ML.NET is currently in Preview. Not all data transforms are currently supported. To submit a request for a certain transform, please open an issue in the [dotnet/machinelearning repository](https://github.com/dotnet/machinelearning/issues).
+
+## Categorical
+
+| Transform | Definition |
+| --- | --- |
+| CategoricalHashOneHotVectorizer | Encodes the categorical variable with hash-based encoding. |
+| CategoricalOneHotVectorizer | Encodes the categorical variable with one-hot encoding based on term dictionary. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: on term dictionary --> on a term dictionary

File: docs/machine-learning/resources/transforms.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+## Combiners and segregators
+
+| Transform | Definition |
+| --- | --- |
+| CombinerByContiguousGroupId | Groups values of a scalar column into a vector, by a contiguous group ID. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: no comma after vector; better: by --> based on

File: docs/machine-learning/resources/transforms.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+| FeatureCombiner | Combines all the features into one feature column. |
+| ManyHeterogeneousModelCombiner | Combines a sequence of TransformModels and a PredictorModel into a single PredictorModel. |
+| ModelCombiner	| Combines a sequence of TransformModels into a single model. |
+| Segregator | Un-groups vector columns into sequences of rows, inverse of Group transform. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: Un-groups --> ungroups
rows, inverse  --> rows; the inverse 
Group transform isn't a transformation listed in this topic; should it be?

File: docs/machine-learning/resources/transforms.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+| Transform | Definition |
+| --- | --- |
+| FeatureSelectorByCount | Selects the slots for which the count of non-default values is greater than or equal to a threshold. |
+| FeatureSelectorByMutualInformation | Selects the top k slots across all specified columns ordered by their mutual information with the label column. |
+
+## Featurizers
+
+| Transform | Definition |
+| --- | --- |
+| HashConverter	| Converts column values into hashes. This transform accepts both numeric and text inputs, both single and vector-valued columns. This is a part of the Dracula transform. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: what is the Dracula transform?

File: docs/machine-learning/resources/transforms.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+| TreeLeafFeaturizer | Trains a tree ensemble, or loads it from a file, then maps a numeric feature vector to three outputs: 1. A vector containing the individual tree outputs of the tree ensemble. 2. A vector indicating the leaves that the feature vector falls on in the tree ensemble. 3. A vector indicating the paths that the feature vector falls on in the tree ensemble. If a both a model file and a trainer are specified - will use the model file. If neither are specified, will train a default FastTree model. This can handle key labels by training a regression model towards their optionally permuted indices. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: If a both a model file --> If both a model file.
there should be a subject that corresponds to the verbs "will use" and "will train"

File: docs/machine-learning/resources/transforms.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+## Label parsing
+
+| Transform | Definition |
+| --- | --- |
+| Dictionarizer	| Converts input values (words, numbers, etc.) to index in a dictionary. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: to index --> to indices (?)

File: docs/machine-learning/resources/transforms.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+| LabelColumnKeyBooleanConverter | Transforms the label to either key or bool (if needed) to make it suitable for classification. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: the label --> a label
key or bool --> a key or a bool

File: docs/machine-learning/resources/transforms.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+| LabelIndicator | Label remapper used by OVA. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: what is OVA?

File: docs/machine-learning/resources/transforms.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+| LabelToFloatConverter	| Transforms the label to float to make it suitable for regression. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: to float --> to a float

File: docs/machine-learning/resources/transforms.md [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+| PredictedLabelColumnOriginalValueConverter | Transforms a predicted label column to its original values, unless it is of type bool. |
+ 
+## Missing Values 
+           
+| Transform | Definition |
+| --- | --- |
+| MissingValueHandler | Handle missing values by replacing them with either the default value or the mean/min/max value (for non-text columns only). An indicator column can optionally be concatenated, if theinput column type is numeric. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: Handle --> Handles
concatenated, if theinput --> concatenated if the input

File: docs/machine-learning/resources/transforms.md [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+| MissingValueIndicator	| Create a boolean output column with the same number of slots as the input column, where the output value is true if the value in the input column is missing. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: Create --> Creates

File: docs/machine-learning/resources/transforms.md [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+| MissingValuesDropper | Removes NAs from vector columns. |
+| MissingValuesRowDropper | Filters out rows that contain missing values. |
+| MissingValueSubstitutor | Create an output column of the same type and size of the input column, where missing values are replaced with either the default value or the mean/min/max value (for non-text columns only). |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: Create --> Creates

File: docs/machine-learning/resources/transforms.md [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+## Normalization
+
+| Transform | Definition |
+| --- | --- |
+| BinNormalizer	| The values are assigned into equidensity bins and a value is mapped to its bin_number/number_of_bins. |
+| ConditionalNormalizer	| Normalize the columns only if needed. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: Normalize --> Normalizes

File: docs/machine-learning/resources/transforms.md [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+| GlobalContrastNormalizer | Performs a global contrast normalization on input values: Y = (s * X - M) / D, where s is a scale, M is mean and D is either L2 norm or standard deviation. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: comma between "mean" and "and"
L2 norm or standard deviation --> the L2 norm or the standard deviation

File: docs/machine-learning/resources/transforms.md [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+## Normalization
+
+| Transform | Definition |
+| --- | --- |
+| BinNormalizer	| The values are assigned into equidensity bins and a value is mapped to its bin_number/number_of_bins. |
+| ConditionalNormalizer	| Normalize the columns only if needed. |
+| GlobalContrastNormalizer | Performs a global contrast normalization on input values: Y = (s * X - M) / D, where s is a scale, M is mean and D is either L2 norm or standard deviation. | 
+| LogMeanVarianceNormalizer | Normalizes the data based on the computed mean and variance of the logarithm of the data. |
+| LpNormalizer | Normalize vectors (rows) individually by rescaling them to unit norm (L2, L1 or LInf). Performs the following operation on a vector X: Y = (X - M) / D, where M is mean and D is either L2 norm, L1 norm or LInf norm. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: Normalize --> Normalizes
unit norm --> the unit norm
L2 norm --> the L2 norm

File: docs/machine-learning/resources/transforms.md [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+| OptionalColumnCreator	| If the source column does not exist after deserialization, create a column with the right type and default values. | 
+
+## Text processing and featurization
+
+| Transform | Definition |
+| --- | --- |
+| CharacterTokenizer | Character-oriented tokenizer where text is considered a sequence of characters. |
+| TextFeaturizer | A transform that turns a collection of text documents into numerical feature vectors. The feature vectors are normalized counts of (word and/or character) ngrams in a given tokenized text. |
+| TextToKeyConverter | Converts input values (words, numbers, etc.) to index in a dictionary. |
+| WordTokenizer	| The input to this transform is text, and the output is a vector of text containing the words (tokens) in the original text. The separator is space, but can be specified as any other character (or multiple characters) if needed. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: is space --> is a space, 
can be specified as any other character (or multiple characters) if needed. --> any other character (or multiple characters) can be specified.

File: docs/machine-learning/resources/transforms.md [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+
+## Miscellaneous
+
+| Transform | Definition |
+| --- | --- |
+| ApproximateBootstrapSampler |	Approximate bootstrap sampling.|
+| BinaryPredictionScoreColumnsRenamer | For binary prediction, it renames the PredictedLabel and Score columns to include the name of the positive class.|
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: it renames --> renames

File: docs/machine-learning/resources/transforms.md [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+| DataCache	| Caches using the specified cache option. |
+| DatasetScorer	| Score a dataset with a predictor model. |
+| DatasetTransformScorer | Score a dataset with a transform model. |
+| NoOperation | Does nothing. |
+| RandomNumberGenerator	| Adds a column with a generated number sequence. |
+| ScoreColumnSelector | Selects only the last score columns and the extra columns specified in the arguments. |
+| Scorer | Turn the predictor model into a transform model. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: Turn --> Turns

File: docs/machine-learning/resources/transforms.md [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+| SentimentAnalyzer | Uses a pretrained sentiment model to score input strings. |
+| TrainTestDatasetSplitter | Split the dataset into train and test sets. |
```
Comment (rpetrusha) [inline docs/machine-learning/resources/transforms.md]: Split --> Splits

**Review by rpetrusha — CHANGES_REQUESTED** [review]
This looks good, @aditidugar. I've left a number of comments and suggested changes.

**Comment by aditidugar** [comment]
Thanks at @rpetrusha for the review! Your comments make sense. Before making changes i'm waiting on @GalOshri to review this for accuracy (I'm not sure that all of these transforms are actually available).

Adding @CESARDELATORRE here too.

File: docs/machine-learning/resources/transforms.md (hunk 4 above)
Comment (JRAlexander) [reply docs/machine-learning/resources/transforms.md]: @GalOshri - Same for Dracula transform - should it be added?

**Review by rpetrusha — APPROVED** [review]
LGTM, although adding some additional detail on the Dracula transform would be helpful.

**Comment by JRAlexander** [comment]
Agreed. Will update as soon as we get that feedback.

**Review by JRAlexander — APPROVED** [review]
LGTM.

File: docs/machine-learning/resources/transforms.md [hunk 20]
```diff
@@ -0,0 +1,131 @@
+# Machine learning transforms
```
Comment (pkulikov) [inline docs/machine-learning/resources/transforms.md]: @JRAlexander this file misses metadata.

**Pull request merged**